shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1340,7 +1389,107 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, var ne
         RecordType.mk(fs)(prov)
       } else TupleType(fields.reverseIterator.mapValues(_.toUpper(noProv)))(prov)
   }
-  
+
+  def getNewVarName(prefix: String, nonValidVars: Set[Var])(implicit raise: Raise): String = {
+    // we check all possibe prefix_num combination, till we found one that is not in the nonValidVars
+    val ints = LazyList.from(1)
+    val result = ints.find(index => {
+      !nonValidVars.contains(Var(prefix + "_" + index))
+    })
+    result match {
+      case Some(index) => 
+        prefix + "_" + index
+      case N => 
+        "ERROR" // unreachable, cause there must be an possible NewVar
```

Comment:
You can call `die` for unreachable branches:
```suggestion
        die // unreachable, cause there must be an possible NewVar
```